### PR TITLE
[KubernetesManifestV0] Promote and Reject actions added

### PR DESCRIPTION
**Pull request by nishubansal**
- **Promote** action - If deployment strategy is '**Canary**' it deletes canary deployment (canary and baseline objects) and deploy input manifests. If deployment strategy is not 'Canary', Error message is thrown.
- **Reject** action - If deployment strategy is '**Canary**' it deletes canary deployment (canary and baseline objects). and if deployment strategy is not 'Canary', Error message is thrown.

- Code of other actions is refactored. Common code is moved to Helper classes.
- **kubectl-object-model** is moved out of **utility-commons**.

### CODE REVIEW

File: Tasks/KubernetesManifestV0/task.json [hunk 1]
```diff
@@ -14,8 +14,8 @@
     "author": "Microsoft Corporation",
     "version": {
         "Major": 0,
-        "Minor": 149,
-        "Patch": 0
+        "Minor": 150,
+        "Patch": 5
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/task.json]: 0

File: Tasks/KubernetesManifestV0/src/utils/utilities.ts [hunk 2]
```diff
@@ -122,3 +163,8 @@ export function isEqual(str1: string, str2: string, stringComparer: StringCompar
         return str1 === str2;
     }
 }
+
+function createInlineArray(str: string | string[]): string {
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/utils/utilities.ts]: This is a duplicate function

File: Tasks/KubernetesManifestV0/task.loc.json [hunk 3]
```diff
@@ -75,7 +75,9 @@
         "deploy": "deploy",
         "patch": "patch",
         "scale": "scale",
-        "delete": "delete"
+        "delete": "delete",
+        "reject": "reject",
+        "promote": "promote"
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/task.loc.json]: Shouldn't the names be `promote canary` and `reject canary`, the names aren't quite obvious for a person who hasn't chosen a deployment strategy

**Comment by ds-ms** [comment]
@nishubansal could you add some L0s?

File: Tasks/KubernetesManifestV0/task.loc.json (hunk 3 above)
Comment (nishubansal) [reply Tasks/KubernetesManifestV0/task.loc.json]: @shashankbarsin to take a look at names.

Comment (shashankbarsin) [reply Tasks/KubernetesManifestV0/task.loc.json]: Let's keep it generic promote and reject. In future we'll add more strategies like blue/greeen

**Comment by ds-ms** [comment]
You need to bump up the version of all the tasks which take a dependency on common.  New check I suppose. That's breaking your CI build.

File: Tasks/KubernetesManifestV0/Strings/resources.resjson/en-US/resources.resjson [hunk 4]
```diff
@@ -1,20 +1,18 @@
...
+  "loc.input.label.kubernetesServiceConnection": "Kubernetes service connection",
+  "loc.input.help.kubernetesServiceConnection": "Select a Kubernetes service connection.",
   "loc.input.label.namespace": "Namespace",
   "loc.input.help.namespace": "Sets the namespace for the commands by using the –namespace flag. If the namespace is not provided, the commands will run in the default namespace.",
   "loc.input.label.action": "Action",
   "loc.input.help.action": "Choose the action to be performed.",
-  "loc.input.label.deploymentStrategy": "Strategy",
-  "loc.input.help.deploymentStrategy": "Strategy for deployment",
+  "loc.input.label.strategy": "Strategy",
+  "loc.input.help.strategy": "Strategy for deployment",
```
Comment (shashankbarsin) [inline Tasks/KubernetesManifestV0/Strings/resources.resjson/en-US/resources.resjson]: Change to "Deployment strategy to be used"

File: Tasks/KubernetesManifestV0/Tests/L0.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        assert(tr.stderr.indexOf("\"nginx-deployment-canary\" not found") != -1, "Canary deployment is not present");
+        assert(tr.stdout.indexOf("nginx-deployment-canary created") != -1, "Canary deployment is created");
+        assert(tr.stdout.indexOf("nginx-deployment-baseline created") != -1, "Baseline deployment is created");
+        assert(tr.stdout.indexOf("deployment \"nginx-deployment-canary\" successfully rolled out") != -1, "Canary deployment is successfully rolled out");
+        assert(tr.stdout.indexOf("deployment \"nginx-deployment-baseline\" successfully rolled out") != -1, "Baseline deployment is successfully rolled out");
+        assert(tr.stdout.indexOf("nginx-deployment-canary annotated") != -1, "Canary deployment is annotated");
+        assert(tr.stdout.indexOf("nginx-deployment-baseline annotated") != -1, "Baseline deployment is annotated");
+        done();
+    });
+    it('Run should fail when canary deployment already exit', (done:MochaDone) => {
```
Comment (shashankbarsin) [inline Tasks/KubernetesManifestV0/Tests/L0.ts]: Change to "exists"

File: Tasks/KubernetesManifestV0/Tests/L0.ts [hunk 6]
```diff
@@ -0,0 +1,133 @@
+import * as path from 'path';
+import * as assert from 'assert';
+import * as ttm from 'vsts-task-lib/mock-test';
+import tl = require('vsts-task-lib');
+import * as shared from './TestShared';
+
+describe('Kubernetes Manifests Suite', function() {
```
Comment (shashankbarsin) [inline Tasks/KubernetesManifestV0/Tests/L0.ts]: We need to add tests for bake, scale and patch actions as well

File: Tasks/KubernetesManifestV0/src/actions/promote.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import * as utils from "../utils/utilities";
+import * as TaskInputParameters from '../models/TaskInputParameters';
+
+export async function promote() {
+
+    let kubectl = new Kubectl(await utils.getKubectl(), TaskInputParameters.namespace);
+
+    // Deploy input manifests
+    tl.debug("Deploying input manifests");
+    deploymentHelper.deploy(kubectl, TaskInputParameters.manifests, "None");
```
Comment (shashankbarsin) [inline Tasks/KubernetesManifestV0/src/actions/promote.ts]: This should be inside the if block for canary strategy check. If one uses action: promote and strategy: none, we should show an error stating that promote action works only with strategy: canary

File: Tasks/Common/utility-common/kubectlutility.ts [hunk 8]
```diff
@@ -9,7 +9,6 @@ const uuidV4 = require('uuid/v4');
 const kubectlToolName = "kubectl";
 export const stableKubectlVersion = "v1.13.2";
```
Comment (shashankbarsin) [inline Tasks/Common/utility-common/kubectlutility.ts]: Change this to "v1.14.0"

**Pull request merged**